quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/integration/resteasy/QuarkusHttpRequest.java

```diff
@@ -0,0 +1,152 @@
+import org.keycloak.quarkus.runtime.integration.jaxrs.EmptyMultivaluedMap;
+import org.keycloak.services.FormPartValueImpl;
+
+import io.vertx.ext.web.RoutingContext;
+
+public final class QuarkusHttpRequest implements HttpRequest {
+
+    private static final MultivaluedMap<String, String> EMPTY_FORM_PARAM = new EmptyMultivaluedMap<>();
+    private static final MultivaluedMap<String, FormPartValue> EMPTY_MULTI_MAP_MULTI_PART = new EmptyMultivaluedMap<>();
+
+    private final ResteasyReactiveRequestContext context;
+
+    public <R> QuarkusHttpRequest(ResteasyReactiveRequestContext context) {
+        this.context = context;
+    }
+
+    @Override
+    public String getHttpMethod() {
+        return context.getMethod();
+    }
+
+    @Override
+    public MultivaluedMap<String, String> getDecodedFormParameters() {
+        FormData parameters = context.getFormData();
+
+        if (parameters == null || !parameters.iterator().hasNext()) {
+            return EMPTY_FORM_PARAM;
+        }
+
+        MultivaluedMap<String, String> params = new QuarkusMultivaluedHashMap<>();
+
+        for (String name : parameters) {
+            Deque<FormValue> values = parameters.get(name);
+
+            if (values == null || values.isEmpty()) {
+                continue;
+            }
+
+            for (FormValue value : values) {
+                params.add(name, value.getValue());
+            }
+        }
+
+        return params;
+    }
+
+    @Override
+    public MultivaluedMap<String, FormPartValue> getMultiPartFormParameters() {
+        FormData formData = context.getFormData();
+
+        if (formData == null) {
+            return EMPTY_MULTI_MAP_MULTI_PART;
+        }
+
+        MultivaluedMap<String, FormPartValue> params = new QuarkusMultivaluedHashMap<>();
+
+        for (String name : formData) {
+            Deque<FormValue> formValues = formData.get(name);
+
+            if (formValues != null) {
+                Iterator<FormValue> iterator = formValues.iterator();
+
+                while (iterator.hasNext()) {
+                    FormValue formValue = iterator.next();
+
+                    if (formValue.isFileItem()) {
+                        try {
+                            params.add(name, new FormPartValueImpl(formValue.getFileItem().getInputStream()));
+                        } catch (IOException cause) {
+                            throw new RuntimeException("Failed to parse multipart file parameter", cause);
+                        }
+                    } else {
+                        params.add(name, new FormPartValueImpl(formValue.getValue()));
+                    }
+                }
+            }
+        }
+
+        return params;
+    }
+
+    @Override
+    public HttpHeaders getHttpHeaders() {
+        return context.getHttpHeaders();
+    }
+
+    @Override
+    public X509Certificate[] getClientCertificateChain() {
+        Instance<RoutingContext> instances = CDI.current().select(RoutingContext.class);
```

Comment:
Can't we directly leverage Arc[1] when it's related directly to Quarkus? Or you wanted to achieve more abstraction via CDI? 

[1] `Arc.container().instance(...`

Reply:
Yes, we can. If not a blocker I would leave it as is.